middlewares/studentAssessmentAccess.js

```diff
@@ -0,0 +1,168 @@
+                                         {maxAge: timeout * 60 * 60 * 1000});
+        if (pwData === null
+            || pwData.password !== res.locals.authz_result.password) {
+            return badPassword(res, 'Password invalid or expired, please try again.');
+        }
+
+        // Successful password case: falls through
+    }
+
+    // Pass-through for everything else
+    next();
+});
+
+router.post('/', function(req, res, next) {
+
+    // For password protected things:
+    if ('authz_result' in res.locals
+        && 'password' in res.locals.authz_result
+        && res.locals.authz_result.password) {
+
+        if (req.body.__action == 'assessmentPassword') {
+
+            if (req.body.password == res.locals.authz_result.password) {
+                var pwCookie = csrf.generateToken({password: req.body.password}, config.secretKey);
+                res.cookie('pl_assessmentpw', pwCookie);
+                return res.redirect(req.originalUrl);
+            } else {
+                return badPassword(res);
+            }
+        }
+    }
+
+    // Fallthrough for the middleware
+    next();
+});
+module.exports = router;
+
+function badPassword(res, msg) {
+
+    logger.info(`invalid password attempt for ${res.locals.user.uid}`);
+    res.clearCookie('pl_assessmentpw');
+    res.locals.passwordMessage = msg;
+    res.locals.prompt = 'password';
+    return res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+}
+
+function badSEB(req, res) {
+
+    var SEBdata = {
+        user_id: res.locals.user.user_id,
+        assessment_id: res.locals.assessment.id,
+        course_instance_id: res.locals.course_instance.id,
+        authz_data: res.locals.authz_data,
+    };
+    res.locals.SEBdata = csrf.generateToken(SEBdata, config.secretKey);
+    var proto = 'seb://';
+    //var proto = 'http://';  // For testing
+    res.locals.SEBUrl = proto + req.get('host') + '/pl/downloadSEBConfig/';
+    res.locals.prompt = 'SEB';
+    return res.status(401).render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+}
+
+function checkUserAgent(res, userAgent) {
+
+    // Check the useragent has SEB in it
+    if (!userAgent.includes('SEB/2')) return;
+
+    // Check user-agent header for exam string
+    var examHash = userAgent.match(/prairielearn:(.+)$/);
+    if (examHash === null) return;
+    var key = examHash[1];
+
+    var fromSEB = csrf.getCheckedData(key, config.secretKey, {maxAge: timeout * 60 * 60 * 1000});
+
+    if ('assessment' in res.locals &&
+        fromSEB.assessment_id == res.locals.assessment.id &&
+        fromSEB.user_id == res.locals.authz_data.user.user_id) {
+
+        res.locals.authz_data.mode = 'SEB';
+    }
+
+    // Assessment list view, enable the mode
+    if (!('assessment' in res.locals)) {
+        res.locals.authz_data.mode = 'SEB';
+    }
+
+    res.locals.authz_data.allowed_assessment_id = fromSEB.assessment_id;
```

Comment:
Where do we check this? I can find any place where `allowed_assessment_id` is used.

Reply:
Unfinished future feature. This middleware triggers off the assessment list page too. This was the start of a way to tell that page only show the allowed assessment but I haven't changed that page.